remove `struct Param` (#143)
abetterinternet/ppm-specification: draft-ppm-protocol.md

* Computing the results of a given measurement
* Reporting results to the collector

We start with some basic type definitions used in other messages.

~~~
enum { prio(0), hits(1) } Proto;

 /* ASCII encoded URL. e.g., "https://example.com" */
opaque Url<1..2^16-1>;

Duration uint64; /* Number of seconds elapsed between two instants */

Time uint64; /* seconds elapsed since start of UNIX epoch */

/* An interval of time, where start is included and end is excluded */
struct {
  Time start;
  Time end;
} Interval;
~~~

## Task Configuration {#task-configuration}

> Prior to the start of the protocol, each endpoint in the system must agree on

Reviewer: ```suggestion
Prior to the start of the protocol, each participant in the protocol must agree on
```

the configuration for each task. A task is uniquely identified by its task ID:

~~~
opaque TaskId[32];
~~~

A `TaskId` is a globally unique sequence of bytes. It is RECOMMENDED that this
be set to a random string output by a cryptographically secure pseudorandom
number generator. Each task has the following parameters associated with it:

> 

Reviewer: Bullets describing parameters with names used elsewhere in the doc should begin with the name. E.g.,:
> * `max_batch_lifetime` is the ...

* aggregator endpoints: A list of URLs relative to which an aggregator's API
  endpoints can be found. Each endpoint's list MUST be in the same order. The

>   leader's endpoint MUST be the first in the list. The order of the

Reviewer: These requirements make me think we should define `struct` for the list of aggregators that defines the order and which endpoint is the leader.

Author: I think you're right, but I think that will come with #138 when we formalize the notion of aggregator roles. Once we have that, I think this language and the discussion of `Report` gets much simpler.

  `encrypted_input_shares` in a `Report` (see {{uploading-reports}}) MUST be the
  same as the order in which aggregators appear in this list.
* collector HPKE configuration: The HPKE configuration of the collector
  (described in {{key-config}}). Having participants agree on this absolves
  collectors of the burden of operating an HTTP server. See
  [#102](https://github.com/abetterinternet/prio-documents/issues/102) for
  discussion. Referred to as `collector_config` elsewhere in this document.
* maximum batch lifetime: The maximum number of times a batch of reports may be
  used in collect requests. Referred to as `max_batch_lifetime` elsewhere in
  this document.
* minimum batch size: The minimum number of reports that appear in a batch.
* minimum batch duration: The minimum time difference between the oldest and
  newest report in a batch. This defines the boundaries with which the batch
  interval of each collect request must be aligned. (See
  {{batch-parameter-validation}}.) Referred to as `min_batch_duration` elsewhere
  i this document.

> * protocol: The PPM protocol, e.g., Prio or Hits.

Reviewer: ```suggestion
* protocol: The core PPM protocol, e.g., Prio or Hits.
```

> * protocol: The PPM protocol, e.g., Prio or Hits.

Reviewer: This is a "named parameter".

* Computing the results of a given measurement
* Reporting results to the collector

We start with some basic type definitions used in other messages.

~~~
enum { prio(0), hits(1) } Proto;

opaque Url<1..2^16-1>; /* ASCII encoded URL. e.g., "https://example.com" */

Duration uint64; /* Number of seconds elapsed between two instants */

Time uint64; /* seconds elapsed since start of UNIX epoch */
~~~

## Task Configuration {#task-configuration}

Prior to the start of the protocol, each endpoint in the system must agree on
the configuration for each task. A task is uniquely identified by its task ID:

~~~
opaque TaskId[32];
~~~

A `TaskId` is a globally unique sequence of bytes. It is RECOMMENDED that this

> be set to a random string derived from a cryptographically secure pseudorandom

Reviewer: ```suggestion
be set to a random string output by a cryptographically secure pseudorandom
```

{{!I-D.irtf-cfrg-hpke}} context for the aggregator by running

~~~
enc, context = SetupBaseS(pk,

>                           "pda input share" || task_id|| server_role)

Reviewer: ```suggestion
                          "pda input share" || task_id || server_role)
```
If this is too long, consider reducing the size of the tab.
